File: Trees/fenwick_tree_max.cpp
#include <bits/stdc++.h>

using namespace std;

// Operations: O(log n)
// Space: O(n)
class FenwickTree {
private:
    vector<int> tree;

    int LSB(int i) {
        return i & -i;
    }

public:
    FenwickTree(int size) {
        tree.resize(size+1, 0);
    }

    // Construct Fenwick Tree
    FenwickTree(const vector<int>& nums) {
        int n = (int)nums.size();
        tree.resize(n+1, 0);

        for(int i = 0; i < n; i++) {
            update(i, nums[i]);
        }
    }

    void update(int index, int value) {
        index++;
        while(index <= (int)tree.size()) {
            tree[index] = max(tree[index], value);
            index += LSB(index);
        }
    }

    // Get prefix sum [0, index]
    int max_query(int index) {
        index++;
        int sum = 0;
        while(index > 0) {
            sum = max(sum, tree[index]);
            index -= LSB(index);
        }
        return sum;
    }

    // Get range sum [l, r] (same as prefix sum formula)
    // int range_sum_query(int l, int r) {
    //     return sum_query(r) - sum_query(l-1);
    // }
};


// testing example
int main(void) {
    vector<int> nums = {1, 3, 5, 7, 9, 11};
    FenwickTree ft(nums);




    return 0;
}

File: legacy/go/ds/fenwick/fenwick_tree_max.go
package fenwick

// Operations: O(log n)
// Space: O(n)
type FenwickTreeMax struct {
	tree []int
}

func (f *FenwickTreeMax) lsb(i int) int {
	return i & -i
}

func NewFenwickTreeMax(size int) *FenwickTreeMax {
	return &FenwickTreeMax{tree: make([]int, size+1)}
}

// Construct Fenwick Tree.
func NewFenwickTreeMaxFromNums(nums []int) *FenwickTreeMax {
	ft := NewFenwickTreeMax(len(nums))
	for i, x := range nums {
		ft.Update(i, x)
	}
	return ft
}

func (f *FenwickTreeMax) Update(index, value int) {
	index++
	for index < len(f.tree) {
		f.tree[index] = max(f.tree[index], value)
		index += f.lsb(index)
	}
}

func (f *FenwickTreeMax) MaxQuery(index int) int {
	index++
	res := 0
	for index > 0 {
		res = max(res, f.tree[index])
		index -= f.lsb(index)
	}
	return res
}
